File: 0015-3sum/0015-3sum.py
class Solution:
    def threeSum(self, nums: list[int]) -> list[list[int]]:
        nums.sort()
        k = 0 
        res = []
        while k<len(nums)-2:
            if k > 0 and nums[k] == nums[k-1]:
                k+=1
                continue
            l = k+1
            r = len(nums) - 1
            while l<r:
                sum = nums[l] + nums[k] + nums[r]
                if sum ==0:
                    res.append([nums[l], nums[k], nums[r]])
                    l+=1
                    r-=1
                    while l<r and nums[l] == nums[l-1]:
                        l+=1
 
                    while l<r and nums[r] == nums[r+1]:
                        r-=1
    
                elif sum > 0:
                    r-=1
                else:
                    l+=1

            k+=1
        return res

        

File: 0015-3sum/0015-3sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        Arrays.sort(nums);

        int k = 0;
        List<List<Integer>> res = new ArrayList<>();

        while(k<nums.length-2){
            if (nums[k] > 0){
                break;
            }
            if (k > 0 && nums[k] == nums[k-1]){
                k+=1;
                continue;
            }

            int l= k+1;
            int r = nums.length-1;

            while(l<r){
                int sum = nums[l] + nums[r] + nums[k];
                if(sum==0){
                    res.add(Arrays.asList(nums[k], nums[l], nums[r]));
                    l+=1;
                    r-=1;

                    while(l<r && nums[l] ==nums[l-1]){
                        l++;
                    }
                    while(l<r && nums[r] == nums[r+1]){
                        r--;
                    }

                }
                else if(sum>0){
                    r-=1;
                }
                else{
                    l+=1;
                }
            }

        k+=1;  
        }
    return res;
    }
}